### PR TITLE
Feature/deployment update

### PR DESCRIPTION
**Pull request by the38th**

### CODE REVIEW

File: core/app/controllers/uffizzi_core/api/cli/v1/projects/deployments_controller.rb [hunk 1]
```diff
@@ -56,6 +56,33 @@ def create
...
+  # @response 401 Not authorized
+  def update
+    compose_file, errors = create_temporary_compose_file
+    return render_invalid_file if compose_file.invalid_file?
+    return render_errors(errors) if errors.present?
+
+    errors = check_credentials(compose_file)
+    return render_errors(errors) if errors.present?
+
+    deployment_id = params[:id]
```
Comment (7R41N33) [inline core/app/controllers/uffizzi_core/api/cli/v1/projects/deployments_controller.rb:80]: We need to pass the deployment object into the service and find it here:

```
Deployment.find(params[:id])
```

File: core/app/controllers/uffizzi_core/api/cli/v1/projects/deployments_controller.rb [hunk 2]
```diff
@@ -56,6 +56,33 @@ def create
...
+    deployment = UffizziCore::DeploymentService.update_from_compose(compose_file, resource_project, current_user, deployment_id)
```
Comment (7R41N33) [inline core/app/controllers/uffizzi_core/api/cli/v1/projects/deployments_controller.rb:81]:
```
updated_deployment = UffizziCore::DeploymentService.update_from_compose(deployment, compose_file, resource_project, current_user)
```

File: core/app/controllers/uffizzi_core/api/cli/v1/projects/deployments_controller.rb [hunk 3]
```diff
@@ -116,6 +135,18 @@ def find_or_create_compose_file
     end
   end
 
+  def create_temporary_compose_file
```
Comment (7R41N33) [inline core/app/controllers/uffizzi_core/api/cli/v1/projects/deployments_controller.rb:138]: Let's move this into a service:

```
UffizziCore::ComposeFileService.create_temporary_compose(
  resource_project,
  current_user,
  compose_file_params,
  dependencies_params[:dependencies],
)
```

File: core/app/services/uffizzi_core/deployment_service.rb [hunk 4]
```diff
@@ -29,6 +29,25 @@ def create_from_compose(compose_file, project, user)
...
+
+      deployment_form = UffizziCore::Api::Cli::V1::Deployment::UpdateForm.new(deployment_attributes)
+      deployment_form.assign_dependences!(project, user)
+      deployment_form.compose_file = compose_file
+
+      if deployment_form.valid?
+        deployment = UffizziCore::Deployment.find(deployment_id)
+        deployment.containers.destroy_all
+        deployment.compose_file.destroy if deployment.compose_file.kind.temporary?
+        deployment.update!(containers: deployment_form.containers, compose_file_id: compose_file.id)
```
Comment (7R41N33) [inline core/app/services/uffizzi_core/deployment_service.rb:43]: You have already invoked validation before this operation. If it does raise an error we need to wrap this logic to a transaction

File: core/app/services/uffizzi_core/deployment_service.rb [hunk 5]
```diff
@@ -29,6 +29,25 @@ def create_from_compose(compose_file, project, user)
...
 
+    def update_from_compose(compose_file, project, user, deployment_id)
+      deployment_attributes = ActionController::Parameters.new(compose_file.template.payload)
+
+      deployment_form = UffizziCore::Api::Cli::V1::Deployment::UpdateForm.new(deployment_attributes)
+      deployment_form.assign_dependences!(project, user)
+      deployment_form.compose_file = compose_file
+
+      if deployment_form.valid?
+        deployment = UffizziCore::Deployment.find(deployment_id)
```
Comment (7R41N33) [inline core/app/services/uffizzi_core/deployment_service.rb:40]: It should be in the controller

File: core/test/controllers/uffizzi_core/api/cli/v1/projects/deployments_controller_test.rb [hunk 6]
```diff
@@ -209,6 +209,62 @@ class UffizziCore::Api::Cli::V1::Projects::DeploymentsControllerTest < ActionCon
...
+    params = {
+      project_slug: @project.slug,
+      compose_file: compose_file_attributes,
+      dependencies: [],
+      id: @deployment[:id],
+    }
+
+    differences = {
+      -> { UffizziCore::ComposeFile.temporary.count } => 1,
+      -> { UffizziCore::Template.with_creation_source(UffizziCore::Template.creation_source.compose_file).count } => 1,
```
Comment (7R41N33) [inline core/test/controllers/uffizzi_core/api/cli/v1/projects/deployments_controller_test.rb:230]: I think we need to add a case where deployment containers are replaced.

File: core/app/controllers/uffizzi_core/api/cli/v1/projects/deployments_controller.rb (hunk 1 above)
Comment (the38th) [reply core/app/controllers/uffizzi_core/api/cli/v1/projects/deployments_controller.rb:80]: resolved in this MR https://github.com/UffizziCloud/uffizzi_app/pull/124

File: core/app/controllers/uffizzi_core/api/cli/v1/projects/deployments_controller.rb (hunk 2 above)
Comment (the38th) [reply core/app/controllers/uffizzi_core/api/cli/v1/projects/deployments_controller.rb:81]: resolved in this MR https://github.com/UffizziCloud/uffizzi_app/pull/124

File: core/app/controllers/uffizzi_core/api/cli/v1/projects/deployments_controller.rb (hunk 3 above)
Comment (the38th) [reply core/app/controllers/uffizzi_core/api/cli/v1/projects/deployments_controller.rb:138]: resolved in this MR https://github.com/UffizziCloud/uffizzi_app/pull/124

File: core/app/services/uffizzi_core/deployment_service.rb (hunk 5 above)
Comment (the38th) [reply core/app/services/uffizzi_core/deployment_service.rb:40]: resolved in this MR https://github.com/UffizziCloud/uffizzi_app/pull/124

File: core/test/controllers/uffizzi_core/api/cli/v1/projects/deployments_controller_test.rb (hunk 6 above)
Comment (the38th) [reply core/test/controllers/uffizzi_core/api/cli/v1/projects/deployments_controller_test.rb:230]: resolved in this MR https://github.com/UffizziCloud/uffizzi_app/pull/124

File: core/app/services/uffizzi_core/deployment_service.rb (hunk 4 above)
Comment (the38th) [reply core/app/services/uffizzi_core/deployment_service.rb:43]: resolved in this MR https://github.com/UffizziCloud/uffizzi_app/pull/124

**Pull request merged**